### DISCUSSION TITLE
Npgsql multiplexing

### DISCUSSION THREAD

**Discussion by pavelvondrasek** [linq2db/linq2db#3075] [Q&A]
Hi,

I have problems on my project with max connection pool because I need to handle thousands of requests and I'm getting errors.

I'm trying Npgsql multiplexing but it gets me some unknown general error messages in stress test. Is it supported or I need another solution to improve pooling?

Thank you!
Pavel

**Comment by pavelvondrasek** [comment]
@MaceWindu Ah, sorry, I didn't know about discussions :)

**Comment by MaceWindu** [comment]
Looking at https://github.com/npgsql/npgsql/issues/1982#issuecomment-756593474 I see that to use it you need:
1. enable multiplexing in connection string
2. use only async APIs

I almost sure it should work fine with both conditions met. If you still have issues it will help to provide more details like error mesage and stack trace

**Reply by pavelvondrasek** [reply]
Hmmm, we tested it on our backend and there only async API calls but it was completely unstable.

Random requests end with 500 :( 

With error messages like:
Error
Exception: Npgsql.NpgsqlException
Message  : Unknown message code: 58
   at Npgsql.Util.PGUtil.ValidateBackendMessageCode(BackendMessageCode code)
   at Npgsql.NpgsqlConnector.ReadMessage(Boolean async, DataRowLoadingMode dataRowLoadingMode, Boolean readingNotifications)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteDbDataReaderAsync(CommandBehavior behavior, CancellationToken cancellationToken)
   at LinqToDB.Data.DataConnection.ExecuteReaderAsync(CommandBehavior commandBehavior, CancellationToken cancellationToken)

All is tracked by Sentry.

**Reply by MaceWindu** [reply]
`Unknown message code` means npgsql receive unexpected data from server which could be:
- bug in npgsql (probably not)
- incorrect use of library, e.g. sharing same connection object (in linq2db case: `DataConnection`) between mutiple threads

I would recommend first to review your code for unsafe use of connections (also make sure you use recent npgsql version).

If it will not help - better to ask for futher directions at https://github.com/npgsql/npgsql

There are a lot of similar reports for this [error message](https://github.com/npgsql/npgsql/issues?q=Unknown+message+code), e.g. https://github.com/npgsql/npgsql/issues/3164

**Reply by MaceWindu** [reply]
Another cause could be missing `await`

**Reply by pavelvondrasek** [reply]
What means unsafe use of connections? I'm using linq2db.aspnet and all connections are provided by DI container.

I use only ToListAsync, ToDictionaryAsync etc. on db context by DI.

Is there another possible case?

**Reply by pavelvondrasek** [reply]
There is Npgsql 5.0.7 and linq2db 3.4 (I updated before test).

**Reply by MaceWindu** [reply]
Well, if you don't share context between DI scopes, then I would recommend to check for missing `await`s (this analyzer could help https://www.nuget.org/packages/Lindhart.Analyser.MissingAwaitWarning/, but it could be slow on big codebase)

**Reply by pavelvondrasek** [reply]
I added analyzer to projects but there still only nullable reference types warnings and I see one warning about async (but its code which isn't used). :(

**Reply by pavelvondrasek** [reply]
<img width="230" alt="Screenshot 2021-06-24 at 23 27 15" src="https://user-images.githubusercontent.com/7081340/123334688-be781d00-d543-11eb-87c5-37ed53427f42.png">

I found that there is one obsolete db context dispose (because old version didnt't use DI)

**Reply by pavelvondrasek** [reply]
Another exceptions are timeouts:

An unhandled exception has occurred while executing the request.
      Npgsql.NpgsqlException (0x80004005): Exception while reading from stream
       ---> System.TimeoutException: Timeout during reading attempt

**Reply by MaceWindu** [reply]
I'm afraid I cannot help you here - it is wether issue in multiplexing implementation by npgsql or with your code.

Could you try to disable multiplexing and check if it still fails? If error is gone - I would recommend to log issue to npgsql repository with sample application that reproduce issue.

**Reply by pavelvondrasek** [reply]
Without multiplexing is app more stable and works great :)

We wanted to use multiplexing because of performance and ability to handle more requests.

So that I will open npgsql issue and I'm able share repository and database dump with them.

**Reply by MaceWindu** [reply]
While looking at this issue https://github.com/npgsql/npgsql/issues/3741, I had another idea for you. it could be that current multiplexing implementation in npgsql could be sensitive to order of secondary operations (like disposal of command, data reader). Could you try to test once more with linq2db `4.0-preview-1` as it implements a bit different logic in this area? Wether it helps or not - please share your repro sample later if it use linq2db so I will help to convert it to raw npgsql calls that linq2db does underneath to allow npgsql team to investigate it.

**Reply by MaceWindu** [reply]
Another thing you can do - try npgsql 6 previews - I see a lot of work on multiplexing in npgsql repo recently and some of those changes were added to 6.0 only (cannot guarantee linq2db will work with 6.0 as we didn't tested it yet)
